apps/browser/settings/declarativeloginmodel.cpp

```diff
@@ -0,0 +1,155 @@
+**
+****************************************************************************/
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "webengine.h"
+
+#include "declarativeloginmodel.h"
+
+static const auto ALL_LOGINS = QStringLiteral("embed:alllogins");
+static const auto LOGINS_ACTION = QStringLiteral("embedui:logins");
+
+DeclarativeLoginModel::DeclarativeLoginModel(QObject *parent) : QAbstractListModel(parent)
+{
+    SailfishOS::WebEngine *webEngine = SailfishOS::WebEngine::instance();
+    webEngine->addObserver(ALL_LOGINS);
+    connect(webEngine, &SailfishOS::WebEngine::recvObserve, this, &DeclarativeLoginModel::handleRecvObserve);
+}
+
+void DeclarativeLoginModel::classBegin()
+{
+}
+
+void DeclarativeLoginModel::componentComplete()
+{
+    requestLogins();
+}
+
+void DeclarativeLoginModel::modify(int index, const QString &username, const QString &password)
+{
+    LoginInfo oldLogin = m_loginList.at(index);
+
+    if ((oldLogin.username() == username) && (oldLogin.password() == password)) {
+        return;
+    }
+
+    LoginInfo newLogin = oldLogin;
+    newLogin.setUsername(username);
+    newLogin.setPassword(password);
+
+    QVariantMap data;
+    data.insert(QStringLiteral("action"), "modify");
+    data.insert(QStringLiteral("oldinfo"), oldLogin.toMap());
+    data.insert(QStringLiteral("newinfo"), newLogin.toMap());
+
+    SailfishOS::WebEngine::instance()->notifyObservers(LOGINS_ACTION, QVariant(data));
+    m_loginList.replace(index, newLogin);
+
+    emit dataChanged(QAbstractListModel::index(index), QAbstractListModel::index(index), QVector<int>());
+}
+
+QVariant DeclarativeLoginModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.row() < 0 || index.row() >= m_loginList.count())
+        return QVariant();
+
+    switch (role) {
+    case Hostname:
+        return m_loginList.at(index.row()).hostname();
+    case Username:
+        return m_loginList.at(index.row()).username();
+    case Password:
+        return m_loginList.at(index.row()).password();
+    default:
+        return QVariant();
+    }
+}
+
+int DeclarativeLoginModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent)
+    return m_loginList.count();
+}
+
+QHash<int, QByteArray> DeclarativeLoginModel::roleNames() const
+{
+    QHash<int, QByteArray> roles;
+    roles[Hostname] = "hostname";
+    roles[Username] = "username";
+    roles[Password] = "password";
+    return roles;
+}
+
+int DeclarativeLoginModel::count() const
+{
+    return rowCount();
+}
+
+void DeclarativeLoginModel::handleRecvObserve(const QString &message, const QVariant &data)
+{
+    if (message == ALL_LOGINS) {
+        setLoginList(qvariant_cast<QVariantList>(data));
+    }
+}
+
+void DeclarativeLoginModel::setLoginList(const QVariantList &data)
```

Comment:
I think if it(https://github.com/sailfishos/sailfish-components-webview/pull/87) approved this code could be used
Taked from https://github.com/sailfishos/sailfish-browser/blob/master/apps/history/declarativehistorymodel.cpp#L113

Reply:
Thanks, that's not a bad idea. However, this case is slightly different to the history model, because while the history model is persistent, this login model is transient, which means that `DeclarativeLoginModel::setLoginList()` will only ever be called on an empty model. As such, although the changes found in `DeclarativeHistoryModel::updateModel()` wouldn't do any harm, they aren't really needed here.

If future changes mean this model becomes persistent, then I think we should make this change.